SurfacePlot Component

syedsalehinipg opened the pull request:
<!-- Insert YouTrack link if relevant -->

Contributes to [TD-2030](https://sce.myjetbrains.com/youtrack/issue/TD-2030)

## Changes

<!-- Let the reviewer know the high-level and detailed changes to look out for -->
<!-- For a bug, this section could instead be bug description & resolution -->

- New **SurfacePlot** component added
- **FullScreenDialog** component added 
  - _to replace the **ConditionalDialog** component which was duplicated in both **LinePlot** and Surface Plot_
- Fixed a bug in **LinePlots** were the title would show twice when in fullscreen mode
- **plotlyConfig.ts** util file added to move reusable function


## UI/UX

<!-- Add in screen grabs / screen recordings of the feature. Tag the UI/UX designer if you require specific feedback / approval -->
<!-- If this PR solves a bug, consider including a before and after so that the reviewer can reproduce and confirm fixed -->

Default component:
![image](https://github.com/IPG-Automotive-UK/react-ui/assets/143519265/b5a3d9d9-228c-4eb3-9dbe-e229263fba72)

![image](https://github.com/IPG-Automotive-UK/react-ui/assets/143519265/9c449960-12ed-46ae-a18a-f858f6c0cc74)

![image](https://github.com/IPG-Automotive-UK/react-ui/assets/143519265/9a27d1c5-9686-4ac3-a3a3-1648d407f416)

In Fullscreen:
![image](https://github.com/IPG-Automotive-UK/react-ui/assets/143519265/38ec8b9d-e816-4656-8a0b-13563abc1b99)

## Testing notes

<!-- Help the reviewer test your feature with some specific steps, point them towards test data and provide scripts or postman configs etc. -->
Test parameters:
`  args: {
    markers: false,
    minHeight: 0,
    showTitle: true,
    title: "Surface Plot",
    xdata: [0.18, 0.36, 0.55, 0.73, 0.91, 1],
    xlabel: "Normalized Torque (-)",
    ydata: [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1],
    ylabel: "Normalized Rotational Speed (-)",
    zdata: [
      [0.76, 0.76, 0.72, 0.7, 0.65, 0.6],
      [0.86, 0.78, 0.76, 0.74, 0.7, 0.65],
      [0.9, 0.9, 0.88, 0.86, 0.86, 0.78],
      [0.92, 0.92, 0.92, 0.9, 0.88, 0.78],
      [0.92, 0.92, 0.92, 0.92, 0.7, 0.65],
      [0.92, 0.94, 0.94, 0.7, 0.65, 0.6],
      [0.92, 0.94, 0.94, 0.7, 0.65, 0.6],
      [0.92, 0.94, 0.7, 0.65, 0.6, 0.6],
      [0.92, 0.94, 0.7, 0.65, 0.6, 0.6],
      [0.92, 0.94, 0.7, 0.65, 0.6, 0.6],
      [0.92, 0.92, 0.7, 0.65, 0.6, 0.6]
    ],
    zlabel: "Efficiency (-)"
  }`

## Author checklist before assigning a reviewer

- [x] Reviewed my own code-diff.
- [ ] Branch has been run in docker.
- [x] PR assigned to me or an appropriate delegate.
- [x] Relevant labels added to the PR.
- [x] Appropriate tests have been added.
- [x] Lint and test workflows pass.

dmbarry86 commented on src/SurfacePlot/SurfacePlot.tsx:
```diff
@@ -0,0 +1,202 @@
+  dialogTitle?: string;
+};
+
+/**
+ * Returns children in a fullscreen dialog if condition is true, otherwise just returns children
+ */
+function ConditionalDialog({
+  condition,
+  onClose,
+  children,
+  dialogTitle
+}: ConditionalDialogProps) {
+  if (condition) {
+    return (
+      <Dialog maxWidth="xl" fullWidth open>
+        <DialogTitle onClose={onClose}>
+          <Box sx={dialogTitle ? {} : { p: 2 }}>{dialogTitle}</Box>
+        </DialogTitle>
+        <DialogContent dividers>
+          <Box style={{ height: "calc(100vh - 168px)" }}>{children}</Box>
+        </DialogContent>
+      </Dialog>
+    );
+  } else {
+    return children;
+  }
+}
```
This code looks identical to the code in LinePlot. Probably it's worth to convert this to it's open component, something like FullscreenDialog or such like.

dmbarry86 commented on src/SurfacePlot/SurfacePlot.tsx:
```diff
@@ -0,0 +1,202 @@
+              t: 50
+            },
+            paper_bgcolor: "rgba(0,0,0,0)",
+            // plot_bgcolor: "rgba(0,0,0,0)",
```
remove commented out code

dmbarry86 commented on src/SurfacePlot/SurfacePlot.tsx:
```diff
@@ -0,0 +1,202 @@
+/**
+ * Returns children in a fullscreen dialog if condition is true, otherwise just returns children
+ */
+function ConditionalDialog({
```
There seems to be a lot of wasted white space around the plot component, particularly so in full screen mode. The plot is very small here, I can't read the axis labels and there is a lot of white space. Please discuss with @Sowbhagya-ipg regarding the size of the plot and margins etc.

![image](https://github.com/IPG-Automotive-UK/react-ui/assets/8976377/bc161019-95d9-4e63-b19c-fccfdbe683f4)

syedsalehinipg replied:
Was unable to remove the white space easily, but I have made improvement to make it larger.

dmbarry86 commented on src/SurfacePlot/SurfacePlot.tsx:
```diff
@@ -0,0 +1,202 @@
+  markers = false,
+  showTitle = false
+}: SurfacePlotProps) => {
+  const theme = useTheme();
```
missing comment

dmbarry86 commented on src/SurfacePlot/SurfacePlot.tsx:
```diff
@@ -0,0 +1,202 @@
+import Plotly from "react-plotly.js";
+import { SurfacePlotProps } from "./SurfacePlot.types";
+
+// Define the SurfacePlot component
```
need a better comment here - these first comments at the top should give a brief description of what the component is used for

dmbarry86 commented on src/SurfacePlot/index.ts:
```diff
@@ -0,0 +1,8 @@
+// This file exports the default export of the SurfacePlotClientOnly module
+// and the SurfacePlotProps type from the SurfacePlot.types module.
```
Don't put sentences on to multiple lines like this. It might look good with your particular width of screen but not for everybody. Just put on one line and let people turn on word wrap in their IDE if they so desire.

dmbarry86 commented on src/LinePlot/LinePlot.tsx:
```diff
@@ -52,7 +52,7 @@ const LinePlot = ({
         width="100%"
       >
-        {showTitle ? (
+        {!isFullscreen && showTitle ? (
```
What's this change here all about? I don't see any mention in the PR description of changes to LinePlot. I assume you have fixed a bug or something but you should mention it in the PR if you touch other components.

syedsalehinipg replied:
Yes it was a bug fix, I have mentioned it in the PR now.

dmbarry86 reviewed (changes requested):
A few suggestions and questions. You should also catch up with @Sowbhagya-ipg so she can provide her approval from UI point of view.

dmbarry86 commented:
@Sowbhagya-ipg are you happy with the UI design? Might be worth a quick call with Syed to get a demo if you haven't already seen.

dmbarry86 commented on src/FullScreenDialog/FullScreenDialog.types.ts:
```diff
@@ -0,0 +1,6 @@
+  condition: boolean;
+  onClose: () => void;
+  children: React.ReactNode;
+  dialogTitle?: string;
```
It would be better to just name this `title` rather than `dialogTitle`. In general we try to stick to single word prop names where possible. There need to be deviations in some cases but in this case title is fine because the componment has dialog in it's name. If it was a button component we wouldn't name the prop buttonColor we would just name it color etc.

Also I think technically this title can actually be any React.ReactNode and it would be rendered.

dmbarry86 commented on src/FullScreenDialog/FullScreenDialog.tsx:
```diff
@@ -0,0 +1,29 @@
+import React from "react";
+
+// This component is used to render a dialog that uses the full width (if the condition is true)
+const FullScreenDialog = (props: FullScreenDialogProps) => {
+  const { condition, onClose, children, dialogTitle } = props;
+  // If the condition is true, render a dialog with the specified title and content
+  if (condition) {
+    return (
+      <Dialog maxWidth="xl" fullWidth open>
+        <DialogTitle onClose={onClose}>
+          <Box sx={dialogTitle ? {} : { p: 2 }}>{dialogTitle}</Box>
+        </DialogTitle>
+        <DialogContent dividers>
+          <Box style={{ height: "calc(100vh - 168px)" }}>{children}</Box>
+        </DialogContent>
+      </Dialog>
+    );
+  }
+  // If the condition is false, render the children as-is
+  else {
+    return children;
+  }
+};
```
I think it would be better to rename this component to `ConditionalDialog`. Now I've seen the code, it's not just a fullscreen dialog, it conditionally shows a dialog or just renders the children directly. Could be misleading to just call it FullScreenDialog.

dmbarry86 commented on src/FullScreenDialog/FullScreenDialog.spec.ts:
```diff
@@ -0,0 +1,24 @@
+import { expect, test } from "@playwright/test";
+
+test("Fullscreen dialog title and content rendered", async ({ page }) => {
+  // Navigate to the FullScreenDialog component in Storybook
+  await page.goto(
+    "http://localhost:6006/?path=/story/dialog-fullscreendialog--default"
+  );
+
+  // Get the title of the fullscreen dialog and assert that it is correct
+  const title = await page
+    .frameLocator('iframe[title="storybook-preview-iframe"]')
+    .getByText("Dialog Title")
+    .textContent();
+  expect(title).toBe("Dialog Title");
+
+  // Get the content of the fullscreen dialog and assert that it is correct
+  const content = await page
+    .frameLocator('iframe[title="storybook-preview-iframe"]')
+    .getByLabel("Dialog Title")
+    .locator("div")
+    .nth(2)
+    .innerHTML();
+  expect(content).toBe("<div>Content goes here</div>");
+});
```
I think we should also test the behaviour where condition is false and therefore we want to check that a dialog does not get rendered.

syedsalehinipg replied:
@dmbarry86 Test added.

dmbarry86 reviewed (changes requested):
A few bits of feedback for the new dialog component and then we're good. I'll leave Sowbhagya to give the UI sign off.

Sowbhagya-ipg commented:
Just had a look at the screenshots of the surfaceplot component. Discussed with @syedsalehinipg on the following issues.
![MicrosoftTeams-image (13)](https://github.com/IPG-Automotive-UK/react-ui/assets/114938041/97f17cdf-a4aa-47b8-8bb5-e6095e94f496)
1. To change the color of the value text-color to white
2. To check the font used in the plot component. It doesn't look like Montserrat
3. To fix the padding around the plot or make the plot larger by default, so it fits around the card with equal space.

VictorCLLucas commented:
One observation from me, we currently have this bug [TD-1949](https://sce.myjetbrains.com/youtrack/issue/TD-1949) which is done using plotly ( I think )we need to fix it in RUI so we don't have the same issue when we start using this component.
I just built RUI locally and we have the same problem.
![image](https://github.com/IPG-Automotive-UK/react-ui/assets/106325764/8a89a9b7-9442-4aa1-9039-492773e518e8)

syedsalehinipg commented:
> Just had a look at the screenshots of the surfaceplot component. Discussed with @syedsalehinipg on the following issues. ![MicrosoftTeams-image (13)](https://user-images.githubusercontent.com/114938041/274546555-97f17cdf-a4aa-47b8-8bb5-e6095e94f496.png)
> 
> 1. To change the color of the value text-color to white
> 2. To check the font used in the plot component. It doesn't look like Montserrat
> 3. To fix the padding around the plot or make the plot larger by default, so it fits around the card with equal space.

- Fonts do seem to be defined as Montserrat
- I have adjusted the colour in dark mode
- I have change the camera angle slightly and adjusted some margins to remove some padding

Normal:
![image](https://github.com/IPG-Automotive-UK/react-ui/assets/143519265/6e38c9ac-c3a5-441c-abcd-f8c9a9ff008b)

Fullscreen:
![image](https://github.com/IPG-Automotive-UK/react-ui/assets/143519265/70e18ac9-9c0e-4f2e-a1d5-b95ab1ebcf90)

Sowbhagya-ipg commented:
Had a UI demo. LGTM

dmbarry86 commented:
From the screenshot you have included there @syedsalehinipg it looks like the vertical colorbar on the right is overlapping the button toolbar you get on hover. Maybe an artefact of resizing the plot?

dmbarry86 commented:
Why do normal and full screen images you included there have different background colors in dark mode? One is black the other grey.

syedsalehinipg commented:
> Why do normal and full screen images you included there have different background colors in dark mode? One is black the other grey.

@dmbarry86 I think this is the normal behavior for storybook, it's to differentiate the difference between the popup dialog and the background.

dmbarry86 commented:
> @dmbarry86 I think this is the normal behavior for storybook, it's to differentiate the difference between the popup dialog and the background.

The component itself has a transparent background then I assume and is just taking the background of the parent container?

syedsalehinipg commented:
> From the screenshot you have included there @syedsalehinipg it looks like the vertical colorbar on the right is overlapping the button toolbar you get on hover. Maybe an artefact of resizing the plot?

I have added a little bit of margin to the top, now it doesn't overlap:

![image](https://github.com/IPG-Automotive-UK/react-ui/assets/143519265/099517de-072f-437f-bd7e-9066a0714a94)

Sowbhagya-ipg commented:
> > Why do normal and full screen images you included there have different background colors in dark mode? One is black the other grey.
> 
> @dmbarry86 I think this is the normal behavior for storybook, it's to differentiate the difference between the popup dialog and the background.

yes this is the way it works, the normal background has a darker background and the pop up has a lighter background

Sowbhagya-ipg commented:
> > From the screenshot you have included there @syedsalehinipg it looks like the vertical colorbar on the right is overlapping the button toolbar you get on hover. Maybe an artefact of resizing the plot?
> 
> I have added a little bit of margin to the top, now it doesn't overlap:
> 
> ![image](https://user-images.githubusercontent.com/143519265/274847464-099517de-072f-437f-bd7e-9066a0714a94.png)

Looks Good to me ! @dmbarry86 @syedsalehinipg

syedsalehinipg commented:
> One observation from me, we currently have this bug [TD-1949](https://sce.myjetbrains.com/youtrack/issue/TD-1949) which is done using plotly ( I think )we need to fix it in RUI so we don't have the same issue when we start using this component. I just built RUI locally and we have the same problem. ![image](https://user-images.githubusercontent.com/106325764/274584642-8a89a9b7-9442-4aa1-9039-492773e518e8.png)

@Sowbhagya-ipg do we need to consider this as a requirement?

Sowbhagya-ipg commented:
> > One observation from me, we currently have this bug [TD-1949](https://sce.myjetbrains.com/youtrack/issue/TD-1949) which is done using plotly ( I think )we need to fix it in RUI so we don't have the same issue when we start using this component. I just built RUI locally and we have the same problem. ![image](https://user-images.githubusercontent.com/106325764/274584642-8a89a9b7-9442-4aa1-9039-492773e518e8.png)
> 
> @Sowbhagya-ipg do we need to consider this as a requirement?
I've reviewed the bug description and the desired behavior mentioned by @JanStehle. In my opinion, displaying very tiny letters isn't ideal, and it doesn't serve a clear purpose. Instead, I propose limiting the extent to which users can zoom in and out on the entire plot to ensure readability.

Following a discussion with @syedsalehinipg , it appears that implementing this change isn't a straightforward task. Therefore, I recommend that we consider making this adjustment in the future but not immediately. @VictorCLLucas

dmbarry86 reviewed (approved):
LGTM. It's nice to get this in RUI.

@evoinea you should make a plan for FLEET/BUILD to start using the LinePlot and SurfacePlot components which are now in RUI.

@syedsalehinipg give me a shout later and we can go through the release process for RUI together as it will be your first time. You'll need to create an account on npm and send me your username so I can add you.
